src/riak_kv_ddl_compiler.erl

```diff
@@ -0,0 +1,878 @@
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% TODO
+%% * are we usings lists to describe fields or what?
+%% * turning bucket names into verification modules is a potential
+%%   memory leak on the atom table - not much we can do about that...
+%% * do we care about the name of the validation module?
+
+%% @doc
+-module(riak_kv_ddl_compiler).
+
+-include("riak_kv_index.hrl").
+-include("riak_kv_ddl.hrl").
+
+-export([
+	 make_validation/1,
+	 make_validation/2
+	]).
+
+-define(NODEBUGOUTPUT, false).
+-define(DEBUGOUTPUT,   true).
+-define(MAKEIGNORE,    true).
+-define(MAKE,          false).
+
+make_validation(#ddl{} = DDL) ->
+    make_validation(DDL, ?NODEBUGOUTPUT, "/tmp").
+
+make_validation(#ddl{} = DDL, OutputDir) ->
+    make_validation(DDL, ?DEBUGOUTPUT, OutputDir).
+
+make_validation(#ddl{} = DDL, HasDebugOutput, OutputDir) ->
+    {Module, AST} = compile(DDL, HasDebugOutput, OutputDir),
+    {ok, Module, Bin} = compile:forms(AST),
+    SpoofBeam = "/tmp/" ++ atom_to_list(Module) ++ ".beam",
+    {module, Module} = code:load_binary(Module, SpoofBeam, Bin).
+
+compile(#ddl{} = DDL, HasDebugOutput, OutputDir) ->
+    #ddl{bucket = Bucket,
+	 fields = Fields} = DDL,
+    {ModName, Attrs, N} = make_attrs(Bucket, 1),
+    {Funs, N2} = recurse_over_fns([Fields], N, 1, [], []),
+    AST = Attrs ++ Funs ++ [{eof, N2}],
+    if HasDebugOutput ->
+	    ASTFileName = filename:join([OutputDir, ModName]) ++ ".ast",
+	    write_file(ASTFileName, io_lib:format("~p", [AST]));
+       el/=se ->
+	    ok
+    end,
+    case erl_lint:module(AST) of
+ 	{ok, []} ->
+     	    if
+     		HasDebugOutput ->
+		    write_src(AST, DDL, ModName, OutputDir);
+     		el/=se ->
+		    ok
+     	    end,
+     	    _Module = {ModName, AST};
+     	Other  -> exit(Other)
+    end.
+
+write_src(AST, DDL, ModName, OutputDir) ->
+    Syntax = erl_syntax:form_list(AST),
+    Header = io_lib:format("%%% Generated Module, do NOT edit~n~n" ++
+			       "Validates the DDL~n~p~n",
+			   [DDL]),
+    Header2 = re:replace(Header, "\\n", "\\\n%%% ", [global, {return, list}]),
+    Src = erl_prettypr:format(Syntax),
+    Contents  = [Header2 ++ "\n" ++ Src],
+    SrcFileName = filename:join([OutputDir, ModName]) ++ ".erl",
+    write_file(SrcFileName, Contents).
+
+write_file(FileName, Contents) ->
+    ok = filelib:ensure_dir(FileName),
+    {ok, Fd} = file:open(FileName, [write]),
+    io:fwrite(Fd, "~s~n", [Contents]),
+    file:close(Fd).
+
+recurse_over_fns([], N, _, [], Acc2) ->
```

Comment:
I think a comment explaining what N and M represent would make this code more understandable. As far as I can tell each AST form needs a unique number, but some numbers seem to be shared among different forms. Also, I'm not exactly sure what M represents in this case. Also, I think the function name could probably be named in a manner describing it's intent and not what it's doing. Maybe something like `build_validation_functions`.